subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/AbstractModuleDependency.java

```diff
@@ -237,7 +247,8 @@ protected ImmutableAttributesFactory getAttributesFactory() {
     }
 
     void setAttributes(AttributeContainerInternal attributes) {
-        this.attributes = attributes;
+        this.attributes = attributes != null ? attributes.asImmutable() : null;
```

Comment:
Having the ability to query an `AttributeContainerInternal` for its mutable / immutable state would be best, as it would allow to set that flag to the right value directly and not have to do potentially costly immutable / mutable conversions.
I would recommend adding `AttributeContainerInternal.isImmutable()` to solve that.

Reply:
So I actually went with that approach first, but it seemed dirtier to require consumers to check this flag first, and potentially forgetting to do that. Maybe doing that makes it cleaner after all, as right now you might set the field directly and wreck the state, since it's all the same class.

FWIW, the only user of this `setAttributes` is in this same class, and guaranteed to pass it an immutable. On a separate note, we could even consider making it private unless you have any objections to that?

Reply:
Ok, I looked deeper at the flow and indeed I would update `setAttributes` to be `private` and call it with an already copied mutable attribute container.

Makes sense?

Reply:
I'm not sure that's always possible.
To copy and make it mutable, you need a factory, but that might be null (technically, I don't know how often this would be the case in practice).
On the other hand, we can always turn a mutable container into an immutable one.

Happy to make setAttributes private though.

Reply:
I suppose you must have had an attribute factory if you managed to set any attributes at all (given `setAttributes` is now private), but we'd still have to check if the current `attributes` is null or not, in order to determine whether to create a new mutable one or not.
And given that the logic is currently copy-on-write (when handling the null -> empty mutable case) I thought it makes more sense to keep it that way.

Reply:
I ended up looking and pushing a change. Let me know if I missed something.